torch/utils/data/dataset.py

```diff
@@ -260,6 +265,125 @@ def __len__(self):
         return len(self.indices)
 
 
+class ChunkDataset(IterableDataset):
+    r"""Dataset which uses hierarchical sampling for efficient data reading.
+
+
+    ``ChunkDataset`` is a stateful dataset that supports hierarchical sampling and
+    efficient reading through chunks. A chunk is selected based on a sampling strategy
+    (first level) and its content is shuffled (second level) before producing
+    a batch or an example.
+
+    In this context, chunks could be files, contents of a folder,
+    sections of large text file, etc. In distributed training, to eliminate different workers reading
+    uneven amount of data, the chunks need to be roughly the same size.
+    Please consult ChunkDataReader for more details.
+
+    ``ChunkDataset`` extends ``IterableDataset`` because the length of the dataset
+    (or simply the size) is unknown. Only the number of chunks is known for the dataset.
+
+    In a distributed setup, each `DataLoader` worker has an instance of ``ChunkDataset`` that uses
+    ``DistributedChunkSampler`` sampler to select chunks based on their global worker rank.
+    Once a chunk index has been selected, it is passed to ``ChunkDataReader`` that will load
+    data into the ``ChunkDataset`` internal cache. Only then ``ChunkDataset`` returns batches.
+
+    Arguments:
+        chunk_sampler (DistributedSampler or DistributedChunkSampler): Draw indices for chunks.
+            Typically used to split data amongst dataloader workers (first level of sampling)
+        chunk_reader (ChunkDataReader): Specialized reader for a given input type
+        shuffle_cache (bool): Setting `True` trigger shuffling of the internal chunk cache
+            (second level of sampling) (default: `True`)
+    """
+
+    def __init__(self, chunk_sampler, chunk_reader, shuffle_cache=True):
+        super(ChunkDataset, self).__init__()
+        assert isinstance(chunk_sampler, DistributedSampler) or isinstance(
+            chunk_sampler, DistributedChunkSampler), 'sampler must be a `DistributedSampler` or `DistributedChunkSampler`'
+        assert callable(chunk_reader), 'chunk_reader must be `callable()` and return a container with data'
+        assert isinstance(shuffle_cache, bool), 'shuffle_cache must be a `bool`'
+
+        self.chunk_sampler = chunk_sampler
```

Comment:
Is there a way to check if chunk_sampler is configured to the total chunk count? Or maybe we enforce a reset for self.chunk_sampler?

The reason I am asking is because if chunk_sampler doesn't reset to chunk count, we may run into some issues that'll be hard to debug later. A check here can spot any potential issues and make it easier for users to use it right.

Reply:
No and yes, respectively. It is not possible to enforce the `chunk_sampler` got the right number of replicas/global worker rank[1], but we can do enforce `reset()` is called before `__next__()`. That is a good idea!

[1] `ChunkDataset` is first instantiated at the main process and at this point, it doesn't know how many workers the `DataLoader` will use (`DataLoader` has access to `ChunkDataset` object, but not the other way around). Once the `ChunkDataset` is cloned by all worker processes, than the number of workers and their worker id become available - that is the context on which `ChunkDataset::reset()` is called to reconfigure `DistributedChunkSampler`

Reply:
I have added `self._is_ready = False` flag at `ChunkDataset` constructor, which is set to `True` at `reset()` method.

If at `__next__`, `reset()` wasnt called already, we call it. This way, a worker initialization function is not strictly necessary... only if the user needs to do something else other than initializing `ChunkDataset`

```
    def __next__(self):
        r"""Returns a batch or raises exception when exhausted"""

        if not self._is_ready:
            self.reset()
```
@jaliyae @xzhu1900

Reply:
With our current design of returning one sample, we will check this condition for each example. I don't think this is good.

Reply:
as per offline discussion, we will keep this approach as it is not expensive